### PR TITLE
Fixing include trace

### PR DESCRIPTION
**Pull request by VMatthijs**
This should hopefully fix issue

https://github.com/stan-dev/stanc3/issues/7

### CODE REVIEW

**Comment by VMatthijs** [comment]
@seantalts , could you maybe have a look at this at some point? (no rush!) This could also be used to produce a good location string for your MIR. I can do that after I have a quick chat with you to know what you'd like it to look like.

File: test/integration/included/stanc.expected [hunk 1]
```diff
@@ -145,20 +175,45 @@ Syntax error at file "sem-err.stan", line 1, columns 2-6, parsing error:
...
+  y ~ normal(0, 1);
+}
+
+
+Warning: deprecated language construct used at file ../included/dep-warning.stan, line 2, column 2, included from
+file ../included/incl_stanc_helper_deprecated_warning.stan, line 2, column 27, included from
+file stanc_helper_with_bad_include_deprecated_warning.stan, line 2, column 50:
+   -------------------------------------------------
+     1:    real z;
+     2:    # I am a deprecated comment.
```
Comment (seantalts) [inline test/integration/included/stanc.expected]: Woo!

File: lib/Errors.ml [hunk 2]
```diff
@@ -145,11 +156,9 @@ let fatal_error ?(msg = "") _ =
 (* Warn that a language feature is deprecated *)
 let warn_deprecated (pos, msg) =
   let file, line, column = position pos in
-  Printf.eprintf
-    "\n\
-     Warning: deprecated language construct used at file \"%s\", line %d, \
-     column %d:\n"
-    file line (column - 1) ;
+  Printf.eprintf "\nWarning: deprecated language construct used at file %s:\n"
+    (append_position_to_filename file
```
Comment (seantalts) [inline lib/Errors.ml]: Seeing a lot of these warning calls, maybe abstract to a function that emits a warning, file name + line nums, and text at the location?

Comment (VMatthijs) [reply lib/Errors.ml]: Do you mean to achieve more code sharing between the different kinds of errors and the warnings? They're all a tiny bit different, but I agree there probably could be more code sharing there.

**Comment by VMatthijs** [comment]
1. If you have a look at the error messages on example models (e.g. in the included folder), then you'll see that you get a full include backtrace on each error or warning. It is still done by storing the include backtrace in the filename, currently.
2. I agree that that would be nicer. I don't know how to pass that state through Menhir, though. What could be done more easily is to convert into one of these data structures after parsing when we construct the AST and then to convert back to a string when we throw an error. That would mean though that the lexing and parsing errors take differently structured location data than the semantic errors. My sense is that I'd rather work on other things at the moment, mostly because I just don't know how to do this at the moment, so it would cost me a lot of time and I'm not sure if the payoff justifies that investment. If you know how to do it in not very much time, please go ahead! I totally agree that it would be the better design.

File: lib/Errors.ml (hunk 2 above)
Comment (seantalts) [reply lib/Errors.ml]: It's related. This is just one specific area where we need to DRY up the code.

**Comment by seantalts** [comment]
I'm not sure I understand your answer to how this adds the information to the AST. Is it stored in the `Location` variant where you've re-purposed the Lexing.position type such that one of its string records now stores additional information?

"pass that state through Menhir" - I think I see what you mean - since the includes are just a rule that replaces the lexbuf with another lexbuf pointing to the new file (and on EOF it just pops one off the global stack). Why not use that global stack (or one like it) to add the position information in the parsing stage?

File: lib/Errors.ml (hunk 2 above)
Comment (VMatthijs) [reply lib/Errors.ml]: What does it my to DRY up the code? Is that an acronym?

**Comment by VMatthijs** [comment]
> I'm not sure I understand your answer to how this adds the information to the AST. Is it stored in the `Location` variant where you've re-purposed the Lexing.position type such that one of its string records now stores additional information?

Yes, the filename together with the include backtrace are stored in the Lexing.position.pos_fname field as a single string. 

> "pass that state through Menhir" - I think I see what you mean - since the includes are just a rule that replaces the lexbuf with another lexbuf pointing to the new file (and on EOF it just pops one off the global stack). Why not use that global stack (or one like it) to add the position information in the parsing stage?
If you know how to do that, please go for it! The only way I know how to get position information out of Menhir for a syntax error or for building the AST is by using `$startpos` and `$endpos`, which gives you the start and end position of a rule (they keep getting updated by Menhir). Of course, you can also manage your own include trace state separately by modifying it in every production, by hand (in the braces after each rule), but that'd have to happen in every single rule and could be ugly and I still wouldn't be 100% sure how to do it: you'd want to get the appropriate include trace for every token and somehow make the include trace for the first and last token a rule matches available.

File: lib/Errors.ml (hunk 2 above)
Comment (seantalts) [reply lib/Errors.ml]: yep - "Don't Repeat Yourself." http://wiki.c2.com/?DontRepeatYourself. People usually moderate that with something like "write everything (twice | thrice | ? ) before abstracting". So it's a balance.

**Comment by seantalts** [comment]
I'm pretty sure that even if we end up not being able to find a way to capture structured location information that we do not want to expose as the AST's permanent location API a hacked Lexing.position :P So it's probably best if we go ahead and do as you suggested and parse out the structured data from the string before inserting it into the AST as something structured - I'm imagining a list/stack of something similar to (but easier to use than) a Lexing.position.

Re: global stack - let me try to understand more. The problem here is that in the parser.mly, you only by default get access to the Lexer's Lexing.position automatically. Here is my question: From the parser, can you not access `lexer.include_stack`?

**Comment by bob-carpenter** [comment]
I like this idea, too.   No need to expose and carry around parser/lexer
internals if a simple stack will suffice (it should given the way includes
work).  It does add some conversion code, though.

**Comment by VMatthijs** [comment]
> I'm pretty sure that even if we end up not being able to find a way to capture structured location information that we do not want to expose as the AST's permanent location API a hacked Lexing.position :P So it's probably best if we go ahead and do as you suggested and parse out the structured data from the string before inserting it into the AST as something structured - I'm imagining a list/stack of something similar to (but easier to use than) a Lexing.position.
> 
> Re: global stack - let me try to understand more. The problem here is that in the parser.mly, you only by default get access to the Lexer's Lexing.position automatically. Here is my question: From the parser, can you not access `lexer.include_stack`?

Sure, if you rewrite the code a bit (i.e. move the include structure to its own file or to parser to avoid a circular dependency between the parser and lexer), you can access `include_stack`, but you'll have to manage it manually if you want to access the include stack of both the start and end position of what a rule matches on. It is easy to just get the include stack at the end of the rule, but note that the start and end might be included from different files. Currently, I'm capturing both the start and end include stack. If we only cared about the include stack at the end of the rule I would've done it this way from the start.

**Comment by VMatthijs** [comment]
@seantalts , do you agree if I merge this and close issue https://github.com/stan-dev/stanc3/issues/7.

Issues https://github.com/stan-dev/stanc3/issues/20 and https://github.com/stan-dev/stanc3/issues/31 will then remain open.

File: lib/Errors.ml (hunk 2 above)
Comment (VMatthijs) [reply lib/Errors.ml]: Cool! I agree that the error messages could be condensed further. Hope to address it in a separate issue: https://github.com/stan-dev/stanc3/issues/31 .

**Comment by seantalts** [comment]
> Sure, if you rewrite the code a bit (i.e. move the include structure to its own file or to parser to avoid a circular dependency between the parser and lexer), 

This seems fine. Is the lexer dependency on the parser just on the tokens? I've see those separated out into their own files pretty often and wasn't sure why - maybe this answers that question.

> you'll have to manage it manually in some clever way while parsing if you want to access the include stack of both the start and end position of what a rule matches on. It is easy to just get the include stack at the end of the rule, but note that the start and end might be included from different files. Currently, I'm capturing both the start and end include stack. If we only cared about the include stack at the end of the rule I would've done it this way from the start.

Good point. Is this important just to print error messages indicating the entire cross-file span correctly? Do you have any example tests illustrating this I could look at?

> Getting the beginning and end include stacks into the AST as a stack would be easy. However, that would still leave us with the issue that this hack remains in the parser and lexer and their error messages.

For me this is the minimum for fixing #20 before launch. The way to think about this is that the AST and MIR data structures are not just the API for huge chunks of the code base, but also for other users who might like to develop tools on top of Stan or contribute additional tools (linters, optimizers, etc). For example, the IBM Research team behind YAPS would like to start using our ocaml parser already likely via the S-expression output and stop maintaining their own ANTLR parser. Generally, even if you hack something together in an ugly way and you decide you want to launch with it, you want to wrap that hack up in a nice interface/API that hides those details.

**Comment by seantalts** [comment]
Re: Merging this - I'm not sure this code is code we will use after fixing #20, but I will merge as it's an improvement and sets the correct standard for include messages.

**Pull request merged**